profiles/killbill/src/main/java/org/killbill/billing/server/notifications/PushNotificationListener.java

```diff
@@ -112,13 +137,104 @@ public Response onCompleted(final Response response) throws Exception {
                         }
                     });
             response = futureStatus.get(timeoutSec, TimeUnit.SECONDS);
+        } catch (final TimeoutException toe) {
+            saveRetryPushNotificationInQueue(tenantId, url, body);
+            return false;
         } catch (final Exception e) {
             log.warn("Failed to push notification url='{}', tenantId='{}'", url, tenantId, e);
             return false;
         }
+
+        if ((response.getStatusCode() == 301) || (response.getStatusCode() == 500)) {
```

Comment:
Agree with @pierre but keep the `saveRetryPushNotificationInQueue` in case of exception. For those there is a question where we should only reinsert on  `TimeoutException` or on all `Exception`. I am tempted to say the later (on all `Exception`).